ocaml/alerts/certificate/certificate_check.ml

```diff
@@ -38,142 +23,73 @@ let get_certificate_attributes rpc session_id =
-  | days, Host _ when days < 8 ->
-      ( cert
-      , Some (body expiring, Api_messages.host_server_certificate_expiring_07)
-      )
-  | days, Host _ when days < 15 ->
-      ( cert
-      , Some (body expiring, Api_messages.host_server_certificate_expiring_14)
-      )
-  | _, Host _ ->
-      ( cert
-      , Some (body expiring, Api_messages.host_server_certificate_expiring_30)
-      )
-  | days, CA _ when days < 0 ->
-      (cert, Some (body expired, Api_messages.pool_ca_certificate_expired))
-  | days, CA _ when days < 8 ->
-      (cert, Some (body expiring, Api_messages.pool_ca_certificate_expiring_07))
-  | days, CA _ when days < 15 ->
-      (cert, Some (body expiring, Api_messages.pool_ca_certificate_expiring_14))
-  | _, CA _ ->
-      (cert, Some (body expiring, Api_messages.pool_ca_certificate_expiring_30))
-  | days, Internal _ when days < 0 ->
-      (cert, Some (body expired, Api_messages.host_internal_certificate_expired))
-  | days, Internal _ when days < 8 ->
-      ( cert
-      , Some (body expiring, Api_messages.host_internal_certificate_expiring_07)
-      )
-  | days, Internal _ when days < 15 ->
-      ( cert
-      , Some (body expiring, Api_messages.host_internal_certificate_expiring_14)
-      )
-  | _, Internal _ ->
-      ( cert
-      , Some (body expiring, Api_messages.host_internal_certificate_expiring_30)
-      )
+let alert_message_cls_and_obj_uuid rpc session_id cert =
+  match cert with
+  | Host (host, _) | Internal (host, _) ->
+      (`Host, XenAPI.Host.get_uuid ~rpc ~session_id ~self:host)
+  | CA (cert, _) ->
+      (`Certificate, XenAPI.Certificate.get_uuid ~rpc ~session_id ~self:cert)
 
-let execute rpc session_id existing_messages (cert, alert) =
-  (* CA-342551: messages need to be deleted if the pending alert regard the
-     same host and has newer, updated information.
-     If the pending alert has the same metadata as the existing host message
-     it must not be emmited and the existing message must be retained, this
-     prevents changing the message UUID.
-     In the case there are alerts regarding the host but no alert is pending
-     they are not destroyed since no alert is automatically dismissed. *)
-  match alert with
-  | Some (body, (alert, priority)) -> (
-    try
-      let cls, obj_uuid =
-        match cert with
-        | Host (host, _) | Internal (host, _) ->
-            (`Host, XenAPI.Host.get_uuid ~rpc ~session_id ~self:host)
-        | CA (cert, _) ->
-            ( `Certificate
-            , XenAPI.Certificate.get_uuid ~rpc ~session_id ~self:cert
-            )
-      in
-      let messages_in_host =
-        List.filter
-          (fun (_, record) -> record.API.message_obj_uuid = obj_uuid)
-          existing_messages
+let get_expiry = function
+  | Host (_, valid_till) | Internal (_, valid_till) | CA (_, valid_till) ->
+      Date.of_float (Date.to_float valid_till +. 1.)
+
+let certificates_to_expiry_message_info_list rpc session_id certificates =
```

Comment:
It seems to me that too many duplicated data are placed in the returned list.

Reply:
Yes, some fields of the same type of certificate will have the same value: `message_cls`, `obj_description`, `message_sent_on_remaining_days_list`, but the others(`message_obj_uuid`, `expiry`) might be different. Is this acceptable? Otherwise, to reduce this duplication, I think the input parameter of the `Expire_alert.alert` will be more complicated.